File: Tree-shaped Deep Forest/Trie/Q3. K-th Smallest in Lexicographical Order/Q3_K-thSmallestinLexicographicalOrder.java
class Solution {

    public int findKthNumber(int n, int k) {
        int curr = 1;          // Start from the smallest lexicographical number
        k--;                   // We already consider '1' as the first number

        // Loop until we reach the k-th number
        while (k > 0) {

            // Count how many numbers exist between current prefix (curr)
            // and the next prefix (curr + 1)
            int step = countSteps(n, curr, curr + 1);

            // If total numbers under this prefix are less than or equal to k
            if (step <= k) {
                // Skip this entire subtree (all numbers starting with curr)
                
                curr++;        // Move to the next prefix (next sibling)
                k -= step;     // Reduce k by how many numbers we skipped

            } else {
                // Otherwise, the k-th number lies inside this subtree
                
                curr *= 10;    // Go deeper into the tree (first child)
                k--;           // We move one step down, so decrease k by 1
            }
        }

        return curr;           // Return the k-th lexicographical number
    }

    // Helper function to count how many numbers exist between prefix1 and prefix2
    private int countSteps(int n, long prefix1, long prefix2) {
        int steps = 0;         // Initialize step counter

        // Continue while prefix1 is within the limit n
        while (prefix1 <= n) {

            // Count numbers between prefix1 and prefix2
            // (bounded by n + 1 to avoid overflow)
            steps += Math.min(n + 1, prefix2) - prefix1;

            // Move to the next level in the tree (add one more digit)
            prefix1 *= 10;
            prefix2 *= 10;
        }

        return steps;          // Return total count of numbers in this range
    }
}

File: Tree-shaped Deep Forest/Trie/Q3. K-th Smallest in Lexicographical Order/Q3_K-thSmallestinLexicographicalOrder.py
class Solution(object):
    def findKthNumber(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        # Start from number 1 (smallest lexicographical number)
        curr = 1
        
        # We already consider '1' as the first number, so reduce k by 1
        k -= 1

        # Loop until we reach the k-th number
        while k > 0:
            # Count how many numbers exist between curr and curr+1 in lexicographical order
            step = self._count_steps(n, curr, curr + 1)

            # If total numbers under this prefix are less than or equal to k
            if step <= k:
                # Skip this entire subtree and move to next prefix
                curr += 1
                k -= step
            else:
                # Go deeper into the subtree (next lexicographical level)
                curr *= 10
                k -= 1  # One step taken downwards

        return curr

    def _count_steps(self, n, prefix1, prefix2):
        """
        Helper function to count numbers between prefix1 and prefix2
        """
        steps = 0

        # Count nodes level by level
        while prefix1 <= n:
            # Add valid numbers between prefix1 and prefix2
            steps += min(n + 1, prefix2) - prefix1
            
            # Move to next level in tree
            prefix1 *= 10
            prefix2 *= 10

        return steps